File: src/auth/utils/validation.ts
import { SignupRequest, SigninRequest } from '../types/auth';

/**
 * Validates email format using standard regex
 */
export function validateEmail(email: string): boolean {
  const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
  return emailRegex.test(email);
}

/**
 * Validates password strength
 * Requirements: minimum 8 characters, at least one uppercase, lowercase, number, and special character
 */
export function validatePassword(password: string): boolean {
  if (password.length < 8) {
    return false;
  }

  // At least one uppercase letter
  const hasUpperCase = /[A-Z]/.test(password);
  // At least one lowercase letter
  const hasLowerCase = /[a-z]/.test(password);
  // At least one digit
  const hasDigit = /\d/.test(password);
  // At least one special character
  const hasSpecialChar = /[!@#$%^&*(),.?":{}|<>]/.test(password);

  return hasUpperCase && hasLowerCase && hasDigit && hasSpecialChar;
}

/**
 * Validates signup request data
 */
export function validateSignupRequest(data: SignupRequest): { isValid: boolean; errors: string[] } {
  const errors: string[] = [];

  if (!data.email) {
    errors.push('Email is required');
  } else if (!validateEmail(data.email)) {
    errors.push('Invalid email format');
  }

  if (!data.password) {
    errors.push('Password is required');
  } else if (!validatePassword(data.password)) {
    errors.push('Password must be at least 8 characters with uppercase, lowercase, number, and special character');
  }

  return {
    isValid: errors.length === 0,
    errors
  };
}

/**
 * Validates signin request data
 */
export function validateSigninRequest(data: SigninRequest): { isValid: boolean; errors: string[] } {
  const errors: string[] = [];

  if (!data.email) {
    errors.push('Email is required');
  } else if (!validateEmail(data.email)) {
    errors.push('Invalid email format');
  }

  if (!data.password) {
    errors.push('Password is required');
  }

  return {
    isValid: errors.length === 0,
    errors
  };
}

File: backend/src/core/validation.py
import re
from typing import Optional


def is_valid_email(email: str) -> bool:
    """
    Validate email format according to standard email format requirements.
    """
    email_regex = r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$'
    return re.match(email_regex, email) is not None


def validate_email_format(email: str) -> tuple[bool, str]:
    """
    Validate email format and return (is_valid, error_message).
    """
    if not email:
        return False, "Email is required"

    if not is_valid_email(email):
        return False, "Invalid email format"

    if len(email) > 254:  # RFC 5321 limit
        return False, "Email is too long"

    return True, ""


def is_valid_password(password: str) -> tuple[bool, str]:
    """
    Validate password strength according to security requirements.
    Returns (is_valid, error_message).
    """
    if len(password) < 8:
        return False, "Password must be at least 8 characters long"

    if not any(c.isupper() for c in password):
        return False, "Password must contain at least one uppercase letter"

    if not any(c.islower() for c in password):
        return False, "Password must contain at least one lowercase letter"

    if not any(c.isdigit() for c in password):
        return False, "Password must contain at least one number"

    if not any(not c.isalnum() for c in password):
        return False, "Password must contain at least one special character"

    return True, ""


def is_valid_name(name: str) -> tuple[bool, str]:
    """
    Validate name format and return (is_valid, error_message).
    """
    if not name:
        return False, "Name is required"

    if len(name) > 100:
        return False, "Name is too long (max 100 characters)"

    if len(name.strip()) == 0:
        return False, "Name cannot be empty or whitespace only"

    return True, ""